### DISCUSSION TITLE
File Browser not finding or placing AWS S3 assets correctly

### DISCUSSION THREAD

**Discussion by AlHolden** [MasaCMS/MasaCMS#83] [General]
We've migrated a collection of site directories from Mura to Masa 7.3.3 (on the same Lucee 5.3.8.206 server). Almost everything works great! Everyone's site assets are hosted on an Amazon S3 instance, and the front-end pages are showing all the S3 imagery just fine. 

FYI
We've made an entry in config/cfapplication.cfm like:
`this.mappings["/s3assets"] = "s3://{access_key}:{secret_key}@/{bucket}/";`
and also these entries in config/settings.ini.cfm:
```
assetdir=/s3assets
assetpath=https://mura-dev.s3.amazonaws.com/
fileAssetPath=${'https://mura-dev.s3.amazonaws.com'}
filedir=/s3assets
```
**The problem is in the File Browser feature.** The _Site Files_ listing there is coming up blank. None of the files in the bucket path appear here for any site. None of the sub folders appear.

We suspect the pathing is wrong, because if we upload a file then the File Browser puts it here... up a level:
https://mura-dev.s3.amazonaws.com/{bucket}/{site}/assets/{uploadedFile}

This is not where Mura had them. They're in folders like this:
https://mura-dev.s3.amazonaws.com/{bucket}/{site}/assets/Image/{should go here}
https://mura-dev.s3.amazonaws.com/{bucket}/{site}/assets/File/
https://mura-dev.s3.amazonaws.com/{bucket}/{site}/assets/Image/
https://mura-dev.s3.amazonaws.com/{bucket}/{site}/assets/Thumbs/

The files are currently accessible from a browser, [like this one](https://mura01.s3.amazonaws.com/aam/assets/Image/unnamed-16.jpg), we have not changed their public status.

**Is there another setting that we missed somewhere**, to get the File Browser hip to the correct place, or display the sub-folders? **Because we copied the site folders over from Mura**, is there something to change in there?
 I searched with all the terms I could think of.

Thanks for any insight!
Al Holden

**Comment by guustnieuwenhuis** [comment]
Hi Al, we're looking into it and get back to you.

**Reply by WaleedEhsan** [reply]
No matter whether S3 is integrated or not, the File Manager is not working. 
https://github.com/MasaCMS/MasaCMS/issues/86

**Comment by AlHolden** [comment]
Below are the results from following these steps:

1. Log in to the admin of a Masa site: https://[CUSTOMER_STAGING_DOMAIN]/admin
2. After the Masa Content panel loads, request the following URL: https://[CUSTOMER_STAGING_DOMAIN]/index.cfm/_api/json/v1/[SITE]/filebrowser/browse

This JSON is returned:
`{"data":{"message":"Permission Denied","permission":{"message":"Permission Denied","success":"0"},"success":"0","dne":"0"},"method":"browse","apiversion":"v1"}`

Here are the response headers:

```
(HTTP/2 200 OK
date: Tue, 17 May 2022 16:58:00 GMT
content-type: application/json;charset=utf-8
strict-transport-security: max-age=1200
generator: Masa CMS 7.3.3
cf-cache-status: DYNAMIC
expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=3v0aAXDch6r%2B70NhOZmpXIxUwIGHTvx0BkHQghqDgqlca0lkjbCNzMZIGAtC9Y50IoTXPjuml4b3QhjvJ4qkXUsb%2B7K%2BzKMFmJJ%2FJI5DMn5fWF3TUdGPK48QKOk9uqO%2Bnrm%2BgVC1wAVKhlCnPKBFcQ%3D%3D"}],"group":"cf-nel","max_age":604800}
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
server: cloudflare
cf-ray: 70cde5371f0f7dbf-LAX
content-encoding: br
X-Firefox-Spdy: h2)
```


The results are the same whether or not Clouflare is in development mode (caching off).

What "permission" is the API looking for here?

**Comment by grantshepert** [comment]
We are looking into a few issues with s3 and the file browser; I will post an update once these are complete.

**Comment by AlHolden** [comment]
Is there anyone we can _PAY_ to resolve this issue? Our whole deployment timeline is kinda blocked due to this one :-]

Alternately, maybe with some hint at which components to look at, we could try to find the cause on our own over here?

Thanks for reading, sorry to kvetch!
Al

**Comment by guustnieuwenhuis** [comment]
Hi @AlHolden, could you reach out to me via email (guust@wearenorth.eu) to schedule a call to discuss the problems you run into?

**Comment by AlHolden** [comment]
I've revised the title of the issue. After some testing, I see that when we upload a file intended for an Amazon S3 bucket (according to the same config we had under Mura) Masa is actually placing the file in a separate path which it created off the server root outside the application.
/assets/[siteid]/assets/
This confusion in correct file location might also explain why the browser window is also opening with nothing to show from the public S3 bucket. The missing COORS headers could be a red herring?

**Comment by AlHolden** [comment]
### Good News

After many trials, **We have a workaround** for this issue that seems to work (although I have a remaining question below).

The **PROBLEM** seems to be based in the legacy Mura instruction [here](https://gist.github.com/stevewithington/49075a04b1220b54805365d51af929e9) to place a mapping line like this in the _/config/cfapplication.cfm_ file:
`this.mappings["/s3assets"] = "s3://{access_key}:{secret_key}@/{your_bucket}/";`

Works in Mura 7.1.476, but **Masa 7.3.8 does not appear to be respecting the above setting**. 
**The fix** is to place a declaration in _/config/settings.ini.cfm_ instead:
`s3assets = s3://{access_key}:{secret_key}@/{your_bucket}/`
I placed this one above the remaining settings which employ "s3assets" in their declarations, just in case.

With this revision, I restarted the Masa instance and the File Browser correctly located, displayed and uploaded to the correct S3 bucket.

My only outstanding question is how to set a "default location" in which the File Browser loads?

- In the Mura version, the {site}/assets/Images folder is first location to load.
- In the Masa version, only the {site}/assets/ loads, and my users now need to be smart enough to click on "Images" before uploading.

Is there a setting for this? The ones in _settings.ini.cfm_ don't seem to change anything.
If there's no immediate answer for this, I can generate a new ticket, as I probably should if I weren't so lazy. ;-]

**Reply by AlHolden** [reply]
To answer my own question above:
There is a `fbFolderTree` cookie which stores the "last" File Browser location visited. I would presume that we can pre-set this at some point outside the core, and the user would also be sent there the "first" time.
`fbFolderTree:"{"foldertree":["Image"]}"`
